### DISCUSSION TITLE
eModbus Bridge - Error E4 - Server ID mismatch

### DISCUSSION THREAD

**Discussion by diegostf** [eModbus/eModbus#58] [eModbus applications]
Hi,

I am testing the WiFi-RTU_Bridge example on ESP32. RTU communication is working.

In another ESP32, I'm running the example TCP03example, requesting information from the first ESP32. However I only get the error: Error E4 - Server ID mismatch.

ESP32 WiFi-RTU_Bridge example, serial out:
. WIFi IP address: 10.0.0.109
[N] 625| ModbusServer.cpp     [ 127] listServer: Server   1:  00
[N] 626| ModbusServer.cpp     [ 127] listServer: Server   2:  00
[N] 627| ModbusServer.cpp     [ 127] listServer: Server   3:  00

In ESP32 TCP03example, request code is:
 Error err = MbTcp.addRequest((uint32_t)millis(), 1, READ_HOLD_REGISTER, 4, 1);

Can you help me resolve this error?

**Comment by bertmelis** [comment]
There's a server id mismatch. You're requesting data from a server with ID `1`. Did you define a server with this ID in the bridge?

**Reply by diegostf** [reply]
In the bridge, the definitions are:

 // ServerID 1, remote server 5
   MBGateway.attachServer(1, 5, ANY_FUNCTION_CODE, &MBRTU);

   // ServerID 2, remote server 6
   MBGateway.attachServer(2, 6, ANY_FUNCTION_CODE, &MBRTU);

   // ServerID 3, remote server 7
   MBGateway.attachServer(3, 7, ANY_FUNCTION_CODE, &MBRTU);


And the Server and ModbusRTU, instances: 

ModbusClientRTU MBRTU(Serial2);  // Create a ModbusRTU client instance
ModbusBridgeWiFi MBGateway(5000);  // Create server

**Reply by Miq1** [reply]
Is there a server with id 5 on your RTU Modbus connected to the bridge? The E4 error code is issued if a server's response has a different server id than that requested.

**Reply by diegostf** [reply]
Yes. I have two devices connected, in ID5 and ID6.

**Comment by Miq1** [comment]
I had a look at the bridge code: you actually found a bug. The bridge's server id is not set in the response received from the remote server.

I will fix that tomorrow.

**Reply by diegostf** [reply]
Thanks for the feedback. I'm waiting to continue testing.

**Comment by Miq1** [comment]
@diegostf : Please get a fresh download of the lib and give it another try - the bug should be fixed now.

**Reply by diegostf** [reply]
Now it's running smooth. Tks @Miq1 .

**Comment by Miq1** [comment]
Thank you for your positive feedback. Have fun!
